lib/lists.nix

```diff
@@ -633,6 +633,42 @@ rec {
     "lib.crossLists is deprecated, use lib.cartesianProductOfSets instead"
     (f: foldl (fs: args: concatMap (f: map f args) fs) [f]);
 
+  /* Remove duplicate adjacent elements from the list, using the supplied
+     equality predicate. O(n) complexity.
+
+     Type: uniqBy :: (a -> a -> bool) -> [a] -> [a]
+
+     Example:
+       uniqBy (x: y: x == y) [ 1 2 2 3 1 1 1 3 1 1 2 ]
+       => [ 1 2 3 1 3 1 2 ]
+  */
+  uniqBy =
+    pred:
+    list:
+    let listLength = length list; in
+    if length list <= 1 then list else
+    let
+      listElemAt = elemAt list;
+      isDupList =
+        genList (n: n != 0 && pred (listElemAt (n - 1)) (listElemAt n)) listLength;
+      isDupAt = elemAt isDupList;
+      uniqLength = foldl' (acc: e: acc + (if e then 0 else 1)) 0 isDupList;
+      uniqNList =
+        let go = n: if isDupAt n then go (n + 1) else n; in
+        genList (n: if n == 0 then 0 else go (uniqNAt (n - 1) + 1)) uniqLength;
+      uniqNAt = elemAt uniqNList;
+    in
+    genList (n: listElemAt (uniqNAt n)) uniqLength;
```

Comment:
This function definitely needs some comments for how it works, but this is seriously impressive how you achieved O(n) like that!

Reply:
Almost all the complexity here has been split out into the `ifilter0` function, so you'll want to review the names & comments there.